// forta-agent/dist/sdk/finding.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Finding = exports.FindingType = exports.FindingSeverity = void 0;
var utils_1 = require("./utils");
var FindingSeverity;
(function (FindingSeverity) {
    FindingSeverity[FindingSeverity["Unknown"] = 0] = "Unknown";
    FindingSeverity[FindingSeverity["Info"] = 1] = "Info";
    FindingSeverity[FindingSeverity["Low"] = 2] = "Low";
    FindingSeverity[FindingSeverity["Medium"] = 3] = "Medium";
    FindingSeverity[FindingSeverity["High"] = 4] = "High";
    FindingSeverity[FindingSeverity["Critical"] = 5] = "Critical";
})(FindingSeverity = exports.FindingSeverity || (exports.FindingSeverity = {}));
var FindingType;
(function (FindingType) {
    FindingType[FindingType["Unknown"] = 0] = "Unknown";
    FindingType[FindingType["Exploit"] = 1] = "Exploit";
    FindingType[FindingType["Suspicious"] = 2] = "Suspicious";
    FindingType[FindingType["Degraded"] = 3] = "Degraded";
    FindingType[FindingType["Info"] = 4] = "Info";
})(FindingType = exports.FindingType || (exports.FindingType = {}));
var Finding = (function () {
    function Finding(name, description, alertId, protocol, severity, type, everestId, metadata) {
        this.name = name;
        this.description = description;
        this.alertId = alertId;
        this.protocol = protocol;
        this.severity = severity;
        this.type = type;
        this.everestId = everestId;
        this.metadata = metadata;
    }
    Finding.prototype.toString = function () {
        return JSON.stringify(__assign(__assign({}, this), { severity: FindingSeverity[this.severity], type: FindingType[this.type] }), null, 2);
    };
    Finding.fromObject = function (_a) {
        var name = _a.name, description = _a.description, alertId = _a.alertId, _b = _a.protocol, protocol = _b === void 0 ? 'ethereum' : _b, severity = _a.severity, type = _a.type, everestId = _a.everestId, _c = _a.metadata, metadata = _c === void 0 ? {} : _c;
        utils_1.assertIsNonEmptyString(name, 'name');
        utils_1.assertIsNonEmptyString(description, 'description');
        utils_1.assertIsNonEmptyString(alertId, 'alertId');
        utils_1.assertIsNonEmptyString(protocol, 'protocol');
        utils_1.assertIsFromEnum(severity, FindingSeverity, 'severity');
        utils_1.assertIsFromEnum(type, FindingType, 'type');
        return new Finding(name, description, alertId, protocol, severity, type, everestId, metadata);
    };
    return Finding;
}());
exports.Finding = Finding;
