// forta-agent/dist/sdk/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keccak256 = exports.ethers = exports.getEthersBatchProvider = exports.getEthersProvider = exports.createBlockEvent = exports.createTransactionEvent = exports.getJsonRpcUrl = exports.Network = exports.EventType = exports.TransactionEvent = exports.BlockEvent = exports.FindingType = exports.FindingSeverity = exports.Finding = void 0;
var ethers_1 = require("ethers");
Object.defineProperty(exports, "ethers", { enumerable: true, get: function () { return ethers_1.ethers; } });
var finding_1 = require("./finding");
Object.defineProperty(exports, "Finding", { enumerable: true, get: function () { return finding_1.Finding; } });
Object.defineProperty(exports, "FindingSeverity", { enumerable: true, get: function () { return finding_1.FindingSeverity; } });
Object.defineProperty(exports, "FindingType", { enumerable: true, get: function () { return finding_1.FindingType; } });
var block_event_1 = require("./block.event");
Object.defineProperty(exports, "BlockEvent", { enumerable: true, get: function () { return block_event_1.BlockEvent; } });
var transaction_event_1 = require("./transaction.event");
Object.defineProperty(exports, "TransactionEvent", { enumerable: true, get: function () { return transaction_event_1.TransactionEvent; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "createBlockEvent", { enumerable: true, get: function () { return utils_1.createBlockEvent; } });
Object.defineProperty(exports, "createTransactionEvent", { enumerable: true, get: function () { return utils_1.createTransactionEvent; } });
Object.defineProperty(exports, "getJsonRpcUrl", { enumerable: true, get: function () { return utils_1.getJsonRpcUrl; } });
Object.defineProperty(exports, "getEthersProvider", { enumerable: true, get: function () { return utils_1.getEthersProvider; } });
Object.defineProperty(exports, "getEthersBatchProvider", { enumerable: true, get: function () { return utils_1.getEthersBatchProvider; } });
Object.defineProperty(exports, "keccak256", { enumerable: true, get: function () { return utils_1.keccak256; } });
var EventType;
(function (EventType) {
    EventType[EventType["BLOCK"] = 0] = "BLOCK";
    EventType[EventType["REORG"] = 1] = "REORG";
})(EventType || (EventType = {}));
exports.EventType = EventType;
var Network;
(function (Network) {
    Network[Network["MAINNET"] = 1] = "MAINNET";
    Network[Network["ROPSTEN"] = 3] = "ROPSTEN";
    Network[Network["RINKEBY"] = 4] = "RINKEBY";
    Network[Network["GOERLI"] = 5] = "GOERLI";
})(Network || (Network = {}));
exports.Network = Network;
