// forta-agent/dist/cli/contracts/agent.registry.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ethers_1 = require("ethers");
var agent_registry_abi_json_1 = __importDefault(require("./agent.registry.abi.json"));
var GAS_MULTIPLIER = 1.15;
var GAS_PRICE_MULTIPLIER = 1.5;
var AgentRegistry = (function () {
    function AgentRegistry(ethersAgentRegistryProvider, agentRegistryContractAddress) {
        this.ethersAgentRegistryProvider = ethersAgentRegistryProvider;
        this.agentRegistryContractAddress = agentRegistryContractAddress;
    }
    AgentRegistry.prototype.agentExists = function (agentId) {
        return __awaiter(this, void 0, void 0, function () {
            var agent;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getContract().getAgent(agentId)];
                    case 1:
                        agent = _a.sent();
                        return [2, !!agent.metadata];
                }
            });
        });
    };
    AgentRegistry.prototype.createAgent = function (fromWallet, agentId, reference) {
        return __awaiter(this, void 0, void 0, function () {
            var from, contract, gas, txOptions, tx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        from = fromWallet.address;
                        contract = this.getContract(fromWallet);
                        return [4, contract.estimateGas.createAgent(agentId, from, reference, [1])];
                    case 1:
                        gas = _a.sent();
                        return [4, this.getTxOptions(gas, fromWallet)];
                    case 2:
                        txOptions = _a.sent();
                        return [4, contract.createAgent(agentId, from, reference, [1], txOptions)];
                    case 3:
                        tx = _a.sent();
                        return [4, tx.wait()];
                    case 4:
                        _a.sent();
                        return [2, tx.hash];
                }
            });
        });
    };
    AgentRegistry.prototype.updateAgent = function (fromWallet, agentId, reference) {
        return __awaiter(this, void 0, void 0, function () {
            var contract, gas, txOptions, tx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        contract = this.getContract(fromWallet);
                        return [4, contract.estimateGas.updateAgent(agentId, reference, [1])];
                    case 1:
                        gas = _a.sent();
                        return [4, this.getTxOptions(gas, fromWallet)];
                    case 2:
                        txOptions = _a.sent();
                        return [4, contract.updateAgent(agentId, reference, [1], txOptions)];
                    case 3:
                        tx = _a.sent();
                        return [4, tx.wait()];
                    case 4:
                        _a.sent();
                        return [2, tx.hash];
                }
            });
        });
    };
    AgentRegistry.prototype.isEnabled = function (agentId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.getContract().isEnabled(agentId)];
            });
        });
    };
    AgentRegistry.prototype.disableAgent = function (fromWallet, agentId) {
        return __awaiter(this, void 0, void 0, function () {
            var contract, gas, txOptions, tx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        contract = this.getContract(fromWallet);
                        return [4, contract.estimateGas.disableAgent(agentId, 1)];
                    case 1:
                        gas = _a.sent();
                        return [4, this.getTxOptions(gas, fromWallet)];
                    case 2:
                        txOptions = _a.sent();
                        return [4, contract.disableAgent(agentId, 1, txOptions)];
                    case 3:
                        tx = _a.sent();
                        return [4, tx.wait()];
                    case 4:
                        _a.sent();
                        return [2, tx.hash];
                }
            });
        });
    };
    AgentRegistry.prototype.enableAgent = function (fromWallet, agentId) {
        return __awaiter(this, void 0, void 0, function () {
            var contract, gas, txOptions, tx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        contract = this.getContract(fromWallet);
                        return [4, contract.estimateGas.enableAgent(agentId, 1)];
                    case 1:
                        gas = _a.sent();
                        return [4, this.getTxOptions(gas, fromWallet)];
                    case 2:
                        txOptions = _a.sent();
                        return [4, contract.enableAgent(agentId, 1, txOptions)];
                    case 3:
                        tx = _a.sent();
                        return [4, tx.wait()];
                    case 4:
                        _a.sent();
                        return [2, tx.hash];
                }
            });
        });
    };
    AgentRegistry.prototype.getContract = function (fromWallet) {
        return new ethers_1.ethers.Contract(this.agentRegistryContractAddress, agent_registry_abi_json_1.default, fromWallet ? fromWallet.connect(this.ethersAgentRegistryProvider) : this.ethersAgentRegistryProvider);
    };
    AgentRegistry.prototype.getTxOptions = function (gasLimit, fromWallet) {
        return __awaiter(this, void 0, void 0, function () {
            var gasPrice;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, fromWallet.connect(this.ethersAgentRegistryProvider).getGasPrice()];
                    case 1:
                        gasPrice = _a.sent();
                        return [2, {
                                gasLimit: Math.round(gasLimit.toNumber() * GAS_MULTIPLIER),
                                gasPrice: Math.round(gasPrice.toNumber() * GAS_PRICE_MULTIPLIER)
                            }];
                }
            });
        });
    };
    return AgentRegistry;
}());
exports.default = AgentRegistry;
